src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.cs

```diff
@@ -498,5 +498,68 @@ private static bool ShouldOptimizeOutInitializer(BoundStatement initializer)
             BoundExpression rhs = assignment.Right;
             return rhs.IsDefaultValue();
         }
+
+        /// <summary>
+        /// Receivers of struct methods are required to be at least RValues but can be assignable variables.
+        /// Whether the mutations from the method are propagated back to the 
+        /// receiver instance is conditional on whether the receiver is a variable that can be assigned. 
+        /// If not, then the invokation is performed on a copy.
+        /// 
+        /// An inconvenient situation may arise when the receiver is an RValue expression (like a ternary operator),
+        /// which is trivially reduced during lowering to one of its operands and 
+        /// such operand happens to be an assignable variable (like a local). That operation alone would 
+        /// expose the operand to mutations while it would not be exposed otherwise.
+        /// I.E. the transformation becomes semantically observable.
+        /// 
+        /// To prevent such situations, we will wrap the operand into a node whose only 
+        /// purpose is to never be an assignable expression.
+        /// </summary>
+        private static BoundExpression EnsureNotAssignableIfUsedAsMethodReceiver(BoundExpression expr)
+        {
+            // Leave as-is where receiver mutations cannot happen.
+            if (!WouldBeAssignableIfUsedAsMethodReceiver(expr))
+            {
+                return expr;
+            }
+
+            return new BoundConversion(
+                expr.Syntax,
+                expr,
+                Conversion.IdentityValue,
+                @checked: false,
+                explicitCastInCode: true,
+                constantValueOpt: null,
+                type: expr.Type)
+            { WasCompilerGenerated = true };
+        }
+
+        internal static bool WouldBeAssignableIfUsedAsMethodReceiver(BoundExpression receiver)
+        {
+            // - reference type receivers are byval
+            // - special value types (int32, Nullable<T>, . .) do not have mutating members
+            if (receiver.Type.IsReferenceType ||
+                receiver.Type.OriginalDefinition.SpecialType != SpecialType.None)
+            {
+                return false;
+            }
+
+            switch (receiver.Kind)
+            {
+                case BoundKind.Parameter:
+                case BoundKind.Local:
+                case BoundKind.ArrayAccess:
+                case BoundKind.ThisReference:
+                case BoundKind.BaseReference:
+                case BoundKind.PointerIndirectionOperator:
```

Comment:
PointerElementAccess  is not an issue since this node does not happen in lowered trees. 
I have added a test, though, just in case.

---
In reply to: [108300256](https://github.com/dotnet/roslyn/pull/18166#discussion_r108300256) [](ancestors = 108300256,108272377,108272054,108252886)